### PR TITLE
Added another button at the top, a last button called Format on the Transform Dialogue

### LINKED ISSUE #9721: Add Format for numbers into the calculator, and into Transform
@lilyclements and @jkmusyoka with our new presentation of tables I was disturbed that James had to go back to Excel, to present the crops stuff to James.

We may need to improve the crops dialog a bit, but my claim here is that we also need to be able to format both numbers and dates nicely, within a data frame, so they go nicely in a table.  Here is an example:

Image

We are in the process of adding formatting in our gt tables.  Here I suggest the additional option of getting the columns looking nice in the data frame.  Then we can choose the columns, already looking nice for the table, either as they are, or with some reshaping, all in the `Describe > Tables > Presentation dialog`.

R has a powerful format command (and formatC and also prettyNum) all in base R, but I think I mainly prefer Format, from DescTools.  And we already make quite a lot of use of that package.  

If you agree, then I suggest we proceed as follows:

a) Add a Format section (using DescTools) to the Dates/Times keyboard for dates, and to the Transform keyboard for numbers.  In each case the resulting columns are of type C.  There's a lot of options, so we have multiple keys for the common options.  Maybe 4 or 8 keys in total.  Then users can combine from those options if they want someting more.
b) We start with that.  Then add a `Format` button to the top of the `Prepare > Column: Numeric > Transform` dialog to do the same thing for number columns.
c) Then maybe have buttons at the top of the `Prepare > Column: Dates > Use Date` dialog.  The existing Use Date would become the first - called `Transform` option and we add a `Format` button to format the dates nicely.

One item I have not found in the DescTools:: Format option that I would like is a good decimal.align option for numeric variables.  (I suggest leaving that for now.)  There is one in gt, so maybe we use it there.  If we want it first, then there is this function in the prettyR package.  We could therefore add the package, or maybe take ove that one function and include it in instatExtras?
There is also StrAlign in DescTools - which is parhaps an indication we could also later enhance our Prepare > Column: Text > Transform dialog?

I also don't have the code for a nice Trace button in the Calculate.  This needs a small @lilyclements input.

### PR DESCRIPTION
**Pull request by bayakarisa**
Partly Fixes #9721 
The structure of this button is very similar to numeric. There are 8 radio buttons. They are as follows:

Decimals
Scientific
Percent
NA
Zero
P-value
Fraction
Align

image

Decimals has an up-down from -10 to 22 with default 3.
Scientific has an up-down from 0 to 22 with default 4
Percent has an nup-down from 0 to 5 with default of 2
NA has a pull down into which you can optionally type what you want. Values in the pull down are na, NA, . (i.e. dot), --, missing 
Zero has a pull down where you can type if needed. In the pull down have 0, dot, -- and .
p-value this is a pull down with default p, also possible are * and p*.
Fractions has a pull down into which you can type a posuituve number. Default is 10 and you also have 20 and 100 there. 
Align has a pull down into which you can type. Values are dot, e, left, center, right.

@rdstern @lilyclements @Ag-Derek I have implemented what was discussed
I'll need help putting the R functions of the buttons

### CODE REVIEW

**Comment by lilyclements** [comment]
@bayakarisa this uses the R function `DescTools::Format`

E.g., if you want to see in R:

```r
?DescTools::Format
x <- pi * 10^(-10:10)

DescTools::Format(x, fmt="%")
```

* `<< variable name >>` - this is the name of the variable in the selector.
There are two options with a column receiver: You can either do 

```
ucrReceiver NAME.SetParameterIsString()
```

or

```
ucrReceiverNAME.SetParameterIsRFunction()
```

@bayakarisa why don't you try them both out, and see what happens in the R code for each one? (Hint: Print to script, and look at the `x = VARIABLE NAME` R code. It should be different depending if you set as string vs set as R function)

I've put in the issue some bits on the R code for the rest of the parameters and their values in this function

**Comment by rdstern** [comment]
@bayakarisa just to add to @lilyclements comments:

a) Here is the "old" dialog:

image

In comparison, you have made the dialog quite a lot wider, just to have the quite spacious radio buttons at the top.  I don't mind it a touch wider than before, but can you make the top radio buttons a bit smaller so it is not much changed from before?

Note that all the options are (I think) already working in the Transform keyboard of the calculator.  This should help you with the appropriate R code as the 2 ways should be consistent.  (And you can check they are working in Transform by trying them also in the calculator.

**Comment by rdstern** [comment]
@bayakarisa I noite nothing has happened here for 2 weeks and there is no reply from you, on your plans.  Should this be passed to someone else?

**Comment by bayakarisa** [comment]
@rdstern 

1. Now only  the controls associated with a specific radio button (e.g., the NUD for decimals) are visible if that radio button is selected. 
Format
2. I need help here with inserting the functionality of the radio buttons e.g.... 

This is the current script for  decimal :

```
Format(.x, digits=3)
data_book$append_to_variables_metadata(data_name="survey", col_names="size1", property="labels", new_val="")
```

It should be like this 
```
size <- data_book$get_columns_from_data(data_name="survey", col_names="size", use_current_filter=FALSE)
size1 <- round(size)
data_book$add_columns_to_data(data_name="survey", col_name="size1", col_data=size1, before=FALSE, adjacent_column="size")

data_book$append_to_variables_metadata(data_name="survey", col_names="size1", property="labels", new_val="")
rm(list=c("size1", "size"))
```

**Comment by lilyclements** [comment]
@bayakarisa 

1. I see you want to have this line:

```r
size <- data_book$get_columns_from_data(data_name="survey", col_names="size", use_current_filter=FALSE)
```

This line can come automatically depending on how you set up the receiver type.  There are two options with a column receiver: You can either do

```r
ucrReceiverNAME.SetParameterIsString()
```

or

```
r
ucrReceiverNAME.SetParameterIsRFunction()
```

@bayakarisa why don't you try them both out, and see what happens in the R code for each one? (Hint: Print to script, and look at the x = VARIABLE NAME R code. It should be different depending if you set as string vs set as R function)


2. You're saying as well that you want to run like this:

```r
size1 <- round(size)
```

See [here](https://github.com/IDEMSInternational/R-Instat/issues/9721#issuecomment-3077988843) for the different options that you want to have. These are all in the function `Format` in the `DescTools` package.

So, for your first option, "decimals", you want to have

```r
size1 <- DescTools::Format(size, digits = VALUE IN NUD)
```

For this, you therefore want to have

a. You create a class, and set the function to be `Format`
b. For your class, you want the package name to be `DescTools`
c. For your class, you have two parameters here: First, `size` is the name of the variable in the receiver (check the parameter name for this)
d. Second parameter is `digits`, with the value in the associated NUD. 
e. You then want to assign this to the ucrSave.

**Comment by bayakarisa** [comment]
@lilyclements I have Implemented what you asked 
Kindly check @berylwaswa 

1. I used the Survey data set  and used size 
Format1
2. Here is the script that is generated it and  uses the DescTools::Format

Format2
3. Here is the result - size1 |
Format3

File: instat/dlgTransform.vb [hunk 1]
```diff
@@ -88,9 +87,16 @@ Public Class dlgTransform
...
+    Private clsGetDataFrameFunction As New RFunction
+    Private clsColumnsFunction As New RFunction
+    Private clsPasteFunction As New RFunction
+    Private clsMutateFunction As New RFunction
+    Private clsAcrossFunction As New RFunction
+    Private clsEverythingFunction As New RFunction
+    Private clsPipeOperator As New ROperator
+    Private clsTildaOperator As New ROperator
+    Private clsAssignOperator As New ROperator
+
```
Comment (lilyclements) [inline instat/dlgTransform.vb:99]: Here, I simply moved lines 109-117 up here so that all of our `As New RFunction`'s are together. This can make it easier for the next developer when looking for bits.

File: instat/dlgTransform.vb [hunk 2]
```diff
@@ -227,16 +223,11 @@ Public Class dlgTransform
         ucrPnlFormatOptions.AddParameterValuesCondition(rdoFraction, "check", "Fraction")
         ucrPnlFormatOptions.AddParameterValuesCondition(rdoAlign, "check", "Align")
 
-        ucrReceiverRank.SetParameter(New RParameter("x", 0))
-        ucrReceiverRank.SetParameterIsRFunction()
-
-
```
Comment (lilyclements) [inline instat/dlgTransform.vb]: You set up ucrReceiverRank already (lines 164-169), so I've removed it from here so we don't have duplicated code

File: instat/dlgTransform.vb [hunk 3]
```diff
@@ -272,8 +263,6 @@ Public Class dlgTransform
         ucrPnlFormatOptions.AddToLinkedControls(UcrInputPvalue, {rdoPvalue}, bNewLinkedAddRemoveParameter:=True, bNewLinkedHideIfParameterMissing:=True)
         ucrPnlFormatOptions.AddToLinkedControls({UcrInputAlignOperations, UcrInputAlignValues}, {rdoAlign}, bNewLinkedAddRemoveParameter:=True, bNewLinkedHideIfParameterMissing:=True)
 
-
-
         ucrPnlNonNegative.AddToLinkedControls(ucrInputPower, {rdoPower}, bNewLinkedHideIfParameterMissing:=True)
```
Comment (lilyclements) [inline instat/dlgTransform.vb:266]: Just tidying up here by removing additional lines that are empty. I agree having a break of 1 line is nice and neat and makes things readable between different parts, but, can you go through and ensure there is never a break greater than 1 throughout?

File: instat/dlgTransform.vb [hunk 4]
```diff
@@ -774,11 +763,7 @@ Public Class dlgTransform
         ucrNudLagPosition.AddAdditionalCodeParameterPair(clsLagColsFunction, New RParameter("lag", 1), iAdditionalPairNo:=1)
         ucrNudSignifDigits.AddAdditionalCodeParameterPair(clsSignifColsFunction, New RParameter("digits", 1), iAdditionalPairNo:=1)
         ucrInputPower.AddAdditionalCodeParameterPair(clsPowerColsOperator, New RParameter("y", 1), iAdditionalPairNo:=1)
-
-        ucrReceiverRank.AddAdditionalCodeParameterPair(clsDescToolsFormatFunction, New RParameter("x", 0), iAdditionalPairNo:=0)
```
Comment (lilyclements) [inline instat/dlgTransform.vb]: See next comment, but in that one, I am referring to this line

File: instat/dlgTransform.vb [hunk 5]
```diff
@@ -802,6 +787,7 @@ Public Class dlgTransform
         ucrReceiverRank.AddAdditionalCodeParameterPair(clsScaleMinColsFunction, New RParameter("x", 0), iAdditionalPairNo:=29)
         ucrReceiverRank.AddAdditionalCodeParameterPair(clsBooleanColsOperator, New RParameter("x", 0), iAdditionalPairNo:=30)
         ucrReceiverRank.AddAdditionalCodeParameterPair(clsIsNAColsFunction, New RParameter("x", 0), iAdditionalPairNo:=31)
+        ucrReceiverRank.AddAdditionalCodeParameterPair(clsDescToolsFormatFunction, New RParameter("x", 0), iAdditionalPairNo:=32)
```
Comment (lilyclements) [inline instat/dlgTransform.vb:790]: You had run this before, but, with one very small issue, which resulted in you writing a huge "If" statement, when it could be done much simpler!

The issue? You put `iAdditionalPairNo` as `0`.

We just want to put it as the next number, and put it with it's "brothers" (the other `ucrReceiverRank.AddAdditionalCodeParameterPair`'s.

We can see on lines 787, we have `iAdditionalPairNo := 29`, 788 is `:=30`, 789 is `:=31`, so, we make this one `:=32`.
(Note that 1-28 are all in here too, but to avoid having to write all of them out, I have just highlighted the previous few!)

File: instat/dlgTransform.vb [hunk 6]
```diff
@@ -833,6 +819,7 @@ Public Class dlgTransform
         ucrSaveNew.AddAdditionalRCode(clsPreviewOperator, iAdditionalPairNo:=13)
         ucrSaveNew.AddAdditionalRCode(clsBooleanOperator, iAdditionalPairNo:=14)
         ucrSaveNew.AddAdditionalRCode(clsIsNAFunction, iAdditionalPairNo:=15)
+        ucrSaveNew.AddAdditionalRCode(clsDescToolsFormatFunction, iAdditionalPairNo:=16)
```
Comment (lilyclements) [inline instat/dlgTransform.vb:822]: Exactly the same as above, but with `ucrSaveNew`. I've just moved your `ucrSaveNew.AddAdditionalRCode(clsDescToolsFormatFunction, ....` line to be with it's brothers
And, I've set `iAdditionalPairNo:=16` since it's the 16th one for`ucrSaveNew`.

File: instat/dlgTransform.vb [hunk 7]
```diff
@@ -858,6 +845,9 @@ Public Class dlgTransform
         ucrChkOmitNA.SetRCode(clsMeanFunction, bReset)
         ucrChkPreview.SetRCode(clsConstantDummyFunction, bReset)
 
+        ' Set up for our "Format" options:
+        ucrNudDecimalPlaces.SetRCode(clsDescToolsFormatFunction, bReset)
+
```
Comment (lilyclements) [inline instat/dlgTransform.vb:850]: Now, I've only set this up for `ucrNudDecimalPlaces`, but you need to go through and set this up for the other controls in the Format which are related :) e..g, `ucrNudScientific`, `ucrNudPercent`, and so forth until `ucrInputAlignOperations`.

File: instat/dlgTransform.vb [hunk 8]
```diff
@@ -935,48 +925,7 @@ Public Class dlgTransform
...
-                    ucrBase.clsRsyntax.ClearCodes()
-                    ucrBase.clsRsyntax.AddToBeforeCodes(clsGetColFunc)
-                    ucrBase.clsRsyntax.AddToBeforeCodes(clsFormatFunc)
-                    ucrBase.clsRsyntax.AddToBeforeCodes(clsAddColFunc)
-                    ucrBase.clsRsyntax.SetBaseRFunction(clsFormatFunc)
-                    ucrBase.clsRsyntax.AddToAfterCodes(clsRemoveLabelsFunction)
-                    ucrBase.clsRsyntax.AddToAfterCodes(clsRmFunc)
-
+                If rdoDecimalFormat.Checked Then ' We can remove this If bit too!
                 ElseIf rdoScientific.Checked Then
```
Comment (lilyclements) [inline instat/dlgTransform.vb:929]: This is the fun bit, because I got to remove code! And you will soon too. But I wanted to explain what I've removed and why, so see above.

But, in short, we don't need any of this code here because it's all automatically set up nicely for us when we run the `SetRCode` sub correctly!

File: instat/dlgTransform.vb [hunk 9]
```diff
@@ -935,48 +925,7 @@ Public Class dlgTransform
...
-                    Dim strDataFrame As String = ucrSelectorForRank.ucrAvailableDataFrames.cboAvailableDataFrames.Text
-                    Dim strColumn As String = ucrReceiverRank.GetVariableNames(bWithQuotes:=True)
-                    Dim strTempVar As String = strColumn.Replace("""", "")
-
-                    Dim clsGetColFunc As New RFunction
-                    clsGetColFunc.SetRCommand("data_book$get_columns_from_data")
-                    clsGetColFunc.AddParameter("data_name", Chr(34) & strDataFrame & Chr(34))
-                    clsGetColFunc.AddParameter("col_names", strColumn)
-                    clsGetColFunc.AddParameter("use_current_filter", "FALSE")
-                    clsGetColFunc.SetAssignTo(strTempVar)
```
Comment (lilyclements) [inline instat/dlgTransform.vb]: clsGetColFunc is set up automatically by:

1. We run `ucrReceiverRank.SetParameterIsRFunction()` in `Initialise`
2. We then link this to our `Format` function in the `SetRCode` where we now run `.AddAdditionalCodeParameterPair` with `:= 32`

Which means, this can alllllll go!

File: instat/dlgTransform.vb [hunk 10]
```diff
@@ -935,48 +925,7 @@ Public Class dlgTransform
...
-
-                    Dim strNewVar As String = ucrSaveNew.GetText
-                    Dim clsFormatFunc As New RFunction
-                    clsFormatFunc.SetPackageName("DescTools")
-                    clsFormatFunc.SetRCommand("Format")
```
Comment (lilyclements) [inline instat/dlgTransform.vb]: We set this first bit up in `SetDefaults` so we don't need to repeat it here

```r
                    clsFormatFunc.SetPackageName("DescTools")
                    clsFormatFunc.SetRCommand("Format")
```

File: instat/dlgTransform.vb [hunk 11]
```diff
@@ -935,48 +925,7 @@ Public Class dlgTransform
...
-                    clsFormatFunc.AddParameter("x", strTempVar, bIncludeArgumentName:=False)
-                    clsFormatFunc.AddParameter("digits", ucrNudDecimalPlaces.GetText)
```
Comment (lilyclements) [inline instat/dlgTransform.vb]: This is set up automatically in the `SetRCode` sub where we have `ucrNudDecimalPlaces.SetRCode(...)`, so we don't need to do it manually here. Neat!

File: instat/dlgTransform.vb [hunk 12]
```diff
@@ -935,48 +925,7 @@ Public Class dlgTransform
...
-                    clsGetColFunc.AddParameter("data_name", Chr(34) & strDataFrame & Chr(34))
-                    clsGetColFunc.AddParameter("col_names", strColumn)
-                    clsGetColFunc.AddParameter("use_current_filter", "FALSE")
-                    clsGetColFunc.SetAssignTo(strTempVar)
-
-                    Dim strNewVar As String = ucrSaveNew.GetText
-                    Dim clsFormatFunc As New RFunction
-                    clsFormatFunc.SetPackageName("DescTools")
-                    clsFormatFunc.SetRCommand("Format")
-                    clsFormatFunc.AddParameter("x", strTempVar, bIncludeArgumentName:=False)
```
Comment (lilyclements) [inline instat/dlgTransform.vb]: This is linked to our Format function in the SetRCode where we now run .AddAdditionalCodeParameterPair with := 32

File: instat/dlgTransform.vb [hunk 13]
```diff
@@ -935,48 +925,7 @@ Public Class dlgTransform
...
-
-
-                    Dim clsAddColFunc As New RFunction
-                    clsAddColFunc.SetRCommand("data_book$add_columns_to_data")
-                    clsAddColFunc.AddParameter("data_name", Chr(34) & strDataFrame & Chr(34))
-                    clsAddColFunc.AddParameter("col_name", Chr(34) & strNewVar & Chr(34))
-                    clsAddColFunc.AddParameter("col_data", strNewVar)
-                    clsAddColFunc.AddParameter("before", "FALSE")
-                    clsAddColFunc.AddParameter("adjacent_column", strColumn)
-
```
Comment (lilyclements) [inline instat/dlgTransform.vb]: This is all done automatically in the ucrSave - the ucrSave knows to do all this already, and so we just simply needed to link it to our `Format` function, which we did when ran `ucrSave.AddAdditionalCodeParameterPair with := 16`

File: instat/dlgTransform.vb [hunk 14]
```diff
@@ -935,48 +925,7 @@ Public Class dlgTransform
...
-                    Dim clsRmFunc As New RFunction
-                    clsRmFunc.SetRCommand("rm")
-                    clsRmFunc.AddParameter("list", "c(" & Chr(34) & strNewVar & Chr(34) & ", " & Chr(34) & strTempVar & Chr(34) & ")", bIncludeArgumentName:=True)
-
-
```
Comment (lilyclements) [inline instat/dlgTransform.vb]: This is always done automatically, so this is never needed

File: instat/dlgTransform.vb [hunk 15]
```diff
@@ -935,48 +925,7 @@ Public Class dlgTransform
...
-                    ucrBase.clsRsyntax.ClearCodes()
-                    ucrBase.clsRsyntax.AddToBeforeCodes(clsGetColFunc)
-                    ucrBase.clsRsyntax.AddToBeforeCodes(clsFormatFunc)
-                    ucrBase.clsRsyntax.AddToBeforeCodes(clsAddColFunc)
-                    ucrBase.clsRsyntax.SetBaseRFunction(clsFormatFunc)
-                    ucrBase.clsRsyntax.AddToAfterCodes(clsRemoveLabelsFunction)
-                    ucrBase.clsRsyntax.AddToAfterCodes(clsRmFunc)
-
```
Comment (lilyclements) [inline instat/dlgTransform.vb]: These bits are good, but, it is always the same regardless of which rdo is checked, so we just move this to the end of our `If rdoFormat` statement

**Review by lilyclements — COMMENTED** [review]
@bayakarisa this is looking good! I wrote a long review, then decided the easiest way was to illustrate in one place what the amendments to do are, and then you can apply them.

This took a while to look through, and so I hope it should take you a while too in a way. It's important to understand these bits, and so don't worry about being fast. Worry about understanding it. If you have questions, then ask.

You need to pull changes into your branch too before continuing otherwise you'll get errors. 

So, I've made changes and outlined them for the `Decimal Places` option, you can now take over and do it for the Significant Values` etc options
(Spoiler alert: For Significant Values, I think you just need to add a `.SetRCode` argument where I mentioned it, and remove the code in the If statement!

**Comment by bayakarisa** [comment]
> @bayakarisa please what is the status on this and is it ready for review
@Ag-Derek I need you to step in here

I am getting the same script for Percent and decimal since the parameter for both is digits 
and fraction too 

1.this is the script when i put decimal size 3

size <- data_book$get_columns_from_data(data_name="survey", col_names="size", use_current_filter=FALSE)
size1 <- DescTools::Format(x=size, na.form="Missing", zero.form="0", align="l")
data_book$add_columns_to_data(data_name="survey", col_name="size1", col_data=size1, before=FALSE, adjacent_column="size")

data_book$append_to_variables_metadata(data_name="survey", col_names="size1", property="labels", new_val="")
rm(list=c("size1", "size"))


and here is the script when i put  fractions of 3 on size

size <- data_book$get_columns_from_data(data_name="survey", col_names="size", use_current_filter=FALSE)
size1 <- DescTools::Format(digits=3, x=size, fmt=3, na.form="Missing", zero.form="0", align="l")
data_book$add_columns_to_data(data_name="survey", col_name="size1", col_data=size1, before=FALSE, adjacent_column="size")

data_book$append_to_variables_metadata(data_name="survey", col_names="size1", property="labels", new_val="")
rm(list=c("size1", "size"))

**Comment by Ag-Derek** [comment]
Looks like the `digits` is being called when you actually do the percent parameter, I have sent you a message on whatsapp let me know when you are available within that time

**Review by lilyclements — CHANGES_REQUESTED** [review]

**Comment by lilyclements** [comment]
@Ag-Derek and @berylwaswa can you review this?

**Review by berylwaswa — CHANGES_REQUESTED** [review]
@bayakarisa the radio buttons are all working fine without any errors.

However, I noticed when I first opened the dialog it looked like the image below

instat_2UbBEXDlkY

The decimal places seem to be set by default to 3. But when I click on reset it's changed to zero. And when I go to the next radio button and go back to the decimal radi button it's set to zero. So maybe check that and make sure when someone opens the dialog for the first time it's always set to zero.

@Ag-Derek please check on your end.

**Comment by bayakarisa** [comment]
@lilyclements , Kindly Check
All minimum developer tests have been completed (OK/TestOK, Reset, UI behaviour, and edge cases). No outstanding issues found. .

**Review by lilyclements — CHANGES_REQUESTED** [review]
@bayakarisa the code is working great! Just a couple of minor items:

Can you rearrange the rdo's so 

1. They're equally spaced apart?
2. Can you also move them to the right a bit so they're not covering the group box line on the left hand side? Currently the box is cut off. 
3. They're all aligned to start at the same point

image

You can fix 1 and 3 easily by selecting them all and clicking these:

For 1 -

image

For 3 -

image

Then for two, just selected them all and nudge them to the right a few pixels together. 


Can you also check the Test OK enabled again? For example, here OK is enabled. It shouldn't be enabled if the rdo is checked and the associated NUD is empty

image

**Review by lilyclements — DISMISSED** [review]

**Review by berylwaswa — COMMENTED** [review]
@bayakarisa it's looking good on my side. @rdstern please do a final review before this is merged. Thanks

**Comment by rdstern** [comment]
@bayakarisa et al, sorry I didn't notice this one.  I have been travelling a lot, but please remind me regularly and more quickly in the future.  Thanks @berylwaswa for the recent reminder.

Now I am finding a problem or two:
a) When I return to the dialog it has not "remembered" the Format button.  It returns each time to Numeric at the top.  That's very confusing, because the controls are so similar!
b) When I return to the Format button it seems to have remembered nothing of its prfevious use.  We will often want it to do the same format on a few columns, so the memory is important.
c) I found the first option, decimal seems to work ok.  But look below for the second, namely scientific:
image

 Instead of scientific, it has given 3 decimals and added a percent sign.  Can you please check and correct.
d) That's as far as I got on the checking.  Please can you also check that the options below scientific, do what they should?

**Comment by bayakarisa** [comment]
@rdstern  Kindly Check
All four points have been fixed:
a) Format tab not remembered on reopen — Fixed. 

b) Format sub-options not remembered — Fixed. 

c) Scientific wrong output — Fixed. Changed fmt from "%" to "e" to correctly produce scientific notation per the DescTools documentation.

d) All other options verified

Reset behaviour was also fixed — I added rdoDecimalFormat.Checked = True in ucrBase_ClickReset so Format sub-options always reset back to Decimal.

**Comment by rdstern** [comment]
@bayakarisa that seems a reasonably good improvement.
image

My first problem, shown above, is that the scientific format worked fine, - but just the first time, when it was set to 4, see above.  Then I set it to 2 and it changed the code.  The first time it was:

yield2 <- DescTools::Format(x=yield, digits=4, fmt="e")

But the second time it was:

yield4 <- DescTools::Format(sci=2, digits=2, x=yield)

and was not scientific.

When I made it back to 4 is gave me: yield5 <- DescTools::Format(x=yield), see below:

image

This comes from trying that option multiple times.  I am not sure whether you did this, in your testing.

Let me check that, now with the other options?

The percent, is odd to me.  I would just like the % sign added.  So please can you do yield/100 in your formula.

And it also only works once.  When you use it a second time it has forgotten it is the percent option.

Please check the others in the same way.

**Comment by bayakarisa** [comment]
@rdstern  All the issues you raised have now been fixed:
**Scientific** — the sci parameter was accumulating across multiple uses causing it to override the correct output. This has been fixed by clearing all parameters before each use. Scientific now consistently generates DescTools::Format(x=yield, digits=VALUE, fmt="e") regardless of how many times it is used.

**Percent** — regarding dividing by 100, the fmt="%" code in DescTools::Format already handles this automatically — it divides by 100 and appends the % sign without needing a manual division. So DescTools::Format(x=yield, digits=2, fmt="%") is the correct and complete formula. The repeated use issue has also been fixed.

**All other options** — the root cause of all the repeated use failures was the same accumulation issue, which has now been fixed across all 8 options. Each option now works consistently across multiple uses without losing its settings.
I have tested all 8 options (Decimal, Scientific, Percent, NA, Zero, P-value, Fraction, Align) multiple times each and confirmed they all generate the correct script consistently.

**Comment by rdstern** [comment]
@bayakarisa apologies.  There is (I am afraid) more to do here.  Good that @berylwaswa reminded me.  It is still "in my court", but my fault that it isn't ready!  I'll be as quickly as I can!

a) Percet is ok, but make the default 0.

Then the first bit to Improve, is the NA.

image

b) Make the drop down a bit wider 
c) Omit the blank bit next to it.
d) It should be possible to type into it, e.g. miss.
e) Add `<blank>`, like that, but if chosen then it puts an emptry string into the command, i.e. it is blank!
f) Note the result has 4 decimals.  Instead include digit=2 in the command you generate.  

g) Then almost the same for zero, roughly b, c, d, e, f as above.

h) For the p-value, it is just the generated code that needs to change.  Please add `digits = 3, eps=0.001` into the command in each of the 3 cases.

i) The fraction is ok.  So finally change Align to `Fractions` (so with an s!) in the dialog  Then the R code is simply  to include fmt="frac".   (The calculator, Transform can be used to check.)

**Comment by bayakarisa** [comment]
@rdstern Kindly check

a) Percent: changed default digits from 2 to 0.

Transform Empty Dialog


b) NA option: widened the dropdown, removed the blank/empty item that sat next to it, made the combo box editable so values like "miss" can be typed in, added a literal "<blank>" item that maps to an empty string in the generated code, and changed the generated digits from 4 to 2.

g) Zero option: applied the same b-f changes (wider dropdown, editable, "<blank>" item mapping to an empty string, digits=2).

Adding Blank


h) P-value: added digits=3 and eps=0.001 to the generated command for all three p-value sub-options (*, p*, p).

i) Renamed **rdoAlign to rdoFractions** and changed its displayed text to "Fractions" . The generated R code is now simply
DescTools::Format(x=col, fmt="frac") with no other parameters.
Removed UcrInputAlignOperations, which is no longer needed since the Fractions option takes no input.

**Comment by berylwaswa** [comment]
@rdstern please review this

**Comment by rdstern** [comment]
@bayakarisa please can you resolve the conflicts

**Review by rdstern — DISMISSED** [review]
@bayakarisa well done, I am approving!
@Vitalis95 over to you.  I hope this will be easy, but I'm not sure.

**Review by Vitalis95 — COMMENTED** [review]
@bayakarisa , great work
some few comments then we merge

File: instat/dlgTransform.vb [hunk 16]
```diff
@@ -746,6 +839,8 @@ Public Class dlgTransform
         ucrInputDivide.SetRCode(clsScaleDivideOperator, bReset)
         ucrInputMultiply.SetRCode(clsScaleMultiplyOperator, bReset)
         ucrChkAddConstant.SetRCode(clsConstantDummyFunction, bReset)
+        ucrPnlNumericOptions.SetRCode(clsNumericDummyFunction, bReset)
+        ucrPnlFormatOptions.SetRCode(clsFormatDummyFunction, bReset)
```
Comment (Vitalis95) [inline instat/dlgTransform.vb:843]: remove this, you are setting rcode twice for ucrPnlFormatOptions

File: instat/dlgTransform.vb [hunk 17]
```diff
@@ -756,72 +851,116 @@ Public Class dlgTransform
...
-            ucrSaveNew.Enabled = True
-        End If
     End Sub
 
     Private Sub ucrBase_ClickReset(sender As Object, e As EventArgs) Handles ucrBase.ClickReset
         SetDefaults()
         SetRCodeForControls(True)
         TestOKEnabled()
         ucrChkEditPreview.Checked = False
+        rdoDecimalFormat.Checked = True
```
Comment (Vitalis95) [inline instat/dlgTransform.vb]: do we need this? Seems you have set dummy function and `rdoDecimalFormat` is the default selected option

File: instat/dlgTransform.vb [hunk 18]
```diff
@@ -756,72 +851,116 @@ Public Class dlgTransform
...
 
-    Private Sub ucrPnlTransformOptions_ControlValueChanged(ucrChangedControl As ucrCore) Handles ucrPnlTransformOptions.ControlValueChanged, ucrPnlNumericOptions.ControlValueChanged, ucrInputLogicalValues.ControlValueChanged,
+    Private Function GetParamValue(ucrDropdown As ucrInput, Optional ucrTextbox As ucrInput = Nothing) As String
+        Dim strValue As String = ucrDropdown.GetText()
+        If ucrTextbox IsNot Nothing AndAlso Not String.IsNullOrEmpty(ucrTextbox.GetText()) Then
+            strValue = ucrTextbox.GetText()
+        End If
+        Return strValue
+    End Function
+
```
Comment (Vitalis95) [inline instat/dlgTransform.vb]: Remove this sub, since you are not calling it anywhere.

File: instat/dlgTransform.vb [hunk 19]
```diff
@@ -997,6 +1150,17 @@ Public Class dlgTransform
...
+        ElseIf rdoFormat.Checked Then
+            AddFormatParameters()
+            If rdoDecimalFormat.Checked Then
+            ElseIf rdoScientific.Checked Then
+            ElseIf rdoPercent.Checked Then
+            ElseIf rdoNA.Checked Then
+            ElseIf rdoZero.Checked Then
+            ElseIf rdoPvalue.Checked Then
+            ElseIf rdoFraction.Checked Then
+            ElseIf rdoFractions.Checked Then
```
Comment (Vitalis95) [inline instat/dlgTransform.vb]: It looks like this `If...ElseIf` block is empty and doesn't have any effect. Can it be removed?

File: instat/dlgTransform.vb [hunk 20]
```diff
@@ -1073,6 +1237,108 @@ Public Class dlgTransform
...
+            clsDescToolsFormatFunction.AddParameter("fmt", Chr(34) & UcrInputPvalue.GetText & Chr(34))
+            clsDescToolsFormatFunction.AddParameter("digits", "3")
+            clsDescToolsFormatFunction.AddParameter("eps", "0.001")
+
+        ElseIf rdoFraction.Checked Then
+            clsFormatDummyFunction.AddParameter("check", "Fraction", iPosition:=0)
+            clsDescToolsFormatFunction.AddParameter("fmt", "function(x) instatExtras::frac_den(x, den=" & UcrNudFraction.GetText & ")")
+
+        ElseIf rdoFractions.Checked Then
+            clsFormatDummyFunction.AddParameter("check", "Fractions", iPosition:=0)
```
Comment (Vitalis95) [inline instat/dlgTransform.vb:1269]: Could we specify `iPosition` explicitly for all newly added `AddParameter` calls in this PR ?

File: instat/dlgTransform.vb [hunk 21]
```diff
@@ -1073,6 +1237,108 @@ Public Class dlgTransform
...
+
+        ElseIf rdoFraction.Checked Then
+            UcrNudFraction.Enabled = True
+
+        ElseIf rdoNA.Checked Then
+            UcrInputNAOperations.Enabled = True
+
+        ElseIf rdoZero.Checked Then
+            UcrInputZeroOperations.Enabled = True
+
```
Comment (Vitalis95) [inline instat/dlgTransform.vb]: Do we still need `UpdateControlStates()`? It looks like all of these controls are already linked in `Initialise` using `AddToLinkedControls`

**Comment by bayakarisa** [comment]
@Vitalis95 I have made the necessary changes to the code you can check now

**Comment by berylwaswa** [comment]
@Vitalis95 please check this, thanks

**Comment by Vitalis95** [comment]
@bayakarisa , the translation check is failing. Please fix it, then we can merge.

**Review by Vitalis95 — APPROVED** [review]

**Review by berylwaswa — APPROVED** [review]
